hedera-node/src/main/java/com/hedera/services/stream/RecordStreamFileWriter.java

```diff
@@ -0,0 +1,466 @@
+		this.logPeriodMs = logPeriodMs;
+		this.signer = signer;
+		this.startWriteAtCompleteWindow = startWriteAtCompleteWindow;
+		this.streamType = streamType;
+		this.streamDigest = MessageDigest.getInstance(DigestType.SHA_384.algorithmName());
+		this.metadataStreamDigest = MessageDigest.getInstance(DigestType.SHA_384.algorithmName());
+	}
+
+	@Override
+	public void addObject(T object) {
+		if (checkIfShouldWriteNewFile(object)) {
+			// if we have a current file,
+			// should write endRunningHash, close current file, and generate signature file
+			closeCurrentAndSign();
+			// start new file
+			startNewFile(object);
+			// write the beginning of new file
+			begin();
+		}
+
+		// if recordStreamFile is null, it means startWriteAtCompleteWindow is true,
+		// and we are still in the first incomplete window, so we don't serialize this object;
+		// so we only serialize the object when stream is not null
+		if (recordStreamFile != null) {
+			consume(object);
+		}
+
+		// update runningHash
+		this.runningHash = object.getRunningHash();
+	}
+
+	/**
+	 * check whether need to start a new file
+	 *
+	 * @param object
+	 * 		the object to be written into file
+	 * @return whether we should start a new File for writing this object
+	 */
+	public boolean checkIfShouldWriteNewFile(T object) {
+		Instant currentConsensusTimestamp = object.getTimestamp();
+		boolean result;
+		if (lastConsensusTimestamp == null && !startWriteAtCompleteWindow) {
+			// this is the first object, we should start writing it to new File
+			result = true;
+		} else if (lastConsensusTimestamp == null) {
+			// this is the first object, we should wait for the first complete window
+			result = false;
+		} else {
+			// if lastConsensusTimestamp and currentConsensusTimestamp are in different periods,
+			// we should start a new file
+			result =
+					getPeriod(lastConsensusTimestamp, logPeriodMs) != getPeriod(currentConsensusTimestamp, logPeriodMs);
+		}
+		// update lastConsensusTimestamp
+		lastConsensusTimestamp = currentConsensusTimestamp;
+		return result;
+	}
+
+	/**
+	 * if recordStreamFile is not null:
+	 * write last runningHash to current file;
+	 * close current file;
+	 * and generate a corresponding signature file
+	 */
+	public void closeCurrentAndSign() {
+		if (recordStreamFile != null) {
+			// generate file name
+			final var firstTxnTimestamp = recordStreamFile.getRecordStreamItems(0).getRecord().getConsensusTimestamp();
+			this.file = new File(
+					generateStreamFilePath(
+							Instant.ofEpochSecond(firstTxnTimestamp.getSeconds(), firstTxnTimestamp.getNanos())));
+			this.fileNameShort = file.getName();
+			if (file.exists() && !file.isDirectory()) {
+				LOG.debug(OBJECT_STREAM.getMarker(), "Stream file already exists {}", fileNameShort);
+			} else {
+				try {
+					// write endRunningHash
+					final var endRunningHash = runningHash.getFutureHash().get();
+					recordStreamFile.setEndObjectRunningHash(toProto(endRunningHash));
```

Comment:
Can you remove the algorithm id and the array length from the hash? This line is also putting this info inside the hash.

Reply:
We do want all protobuf fields populated. But we should not calculate any hashes using magic numbers that are not contained within the protobuf fields like algorithm id.